source/simulator/core.cc

```diff
@@ -1889,91 +1891,127 @@ namespace aspect
+            }
 
-        case NonlinearSolver::single_Advection_iterated_Newton_Stokes:
-        {
-          solve_single_advection_and_iterated_newton_stokes(/*use_newton_iterations =*/ true);
-          break;
-        }
+            case NonlinearSolver::single_Advection_iterated_Newton_Stokes:
+            {
+              solve_single_advection_and_iterated_newton_stokes(/*use_newton_iterations =*/ true);
+              break;
+            }
 
-        case NonlinearSolver::single_Advection_no_Stokes:
-        {
-          solve_single_advection_no_stokes();
-          break;
-        }
+            case NonlinearSolver::single_Advection_no_Stokes:
+            {
+              solve_single_advection_no_stokes();
+              break;
+            }
 
-        case NonlinearSolver::first_timestep_only_single_Stokes:
-        {
-          solve_first_timestep_only_single_stokes();
-          break;
-        }
+            case NonlinearSolver::first_timestep_only_single_Stokes:
+            {
+              solve_first_timestep_only_single_stokes();
+              break;
+            }
 
-        case NonlinearSolver::no_Advection_no_Stokes:
-        {
-          solve_no_advection_no_stokes();
-          break;
-        }
+            case NonlinearSolver::no_Advection_no_Stokes:
+            {
+              solve_no_advection_no_stokes();
+              break;
+            }
 
-        default:
-          Assert (false, ExcNotImplemented());
+            default:
+              Assert (false, ExcNotImplemented());
+          }
+        pcout << std::endl;
       }
+    catch (ExcNonlinearSolverNoConvergence &)
+      {
+        pcout << "\nWARNING: The nonlinear solver in the current timestep failed to converge." << std::endl
+              << "Acting according to the parameter 'Nonlinear solver failure strategy'..." << std::endl;
+        ++nonlinear_solver_failures;
 
-    pcout << std::endl;
+        switch (parameters.nonlinear_solver_failure_strategy)
+          {
+            case Parameters<dim>::NonlinearSolverFailureStrategy::continue_with_next_timestep:
+            {
+              pcout << "Continuing to the next timestep even though solution is not fully converged." << std::endl;
+              // do nothing and continue
+              break;
+            }
+            case Parameters<dim>::NonlinearSolverFailureStrategy::cut_timestep_size:
+            {
+              if (timestep_number == 0)
+                {
+                  pcout << "Error: We can not cut the timestep in step 0, so we are aborting."
+                        << std::endl;
+                  throw;
+                }
+              time_stepping_manager.template get_matching_active_plugin<TimeStepping::RepeatOnNonlinearFail<dim>>().nonlinear_solver_has_failed();
+              break;
+            }
+            case Parameters<dim>::NonlinearSolverFailureStrategy::abort_program:
+            {
+              // rethrow the current exception
```

Comment:
the other two options have an additional line that says what is happening next (continue, cut), but this one doesnt, it just throws the error. Can you add another line here, saying that we are now aborting the model run? This would explain to the user we purposefully abort (they would see that from the error message as well, but this makes it more obvious).